cedar-policy-validator/src/typecheck/test/tags.rs

```diff
@@ -0,0 +1,315 @@
+            .exactly_one_underline(r#"principal.getTag(resource.foo).contains("bar")"#)
+            .build(),
+    );
+
+    // works for one principal type this action applies to, but not for all
+    let src = r#"
+        permit(principal, action == Action::"A3", resource) when {
+            principal.hasTag("foo") && principal.getTag("foo") == "bar"
+        };
+    "#;
+    let policy = parse_policy(Some(PolicyID::from_string("0")), src).unwrap();
+    let errors = assert_policy_typecheck_fails(schema_with_tags(), policy);
+    let error = assert_exactly_one_diagnostic(errors);
+    expect_err(
+        src,
+        &miette::Report::new(error),
+        &ExpectedErrorMessageBuilder::error(r#"the types String and Set<String> are not compatible"#)
+            .help("for policy `0`, both operands to a `==` expression must have compatible types. Types must be exactly equal to be compatible")
+            .exactly_one_underline(r#"principal.getTag("foo") == "bar""#)
+            .build(),
+    );
+
+    // works for one action this policy applies to, but not for all
+    let src = r#"
+        permit(principal, action in [Action::"A1", Action::"A2"], resource) when {
+            principal.hasTag("foo") && principal.getTag("foo") == "bar"
+        };
+    "#;
+    let policy = parse_policy(Some(PolicyID::from_string("0")), src).unwrap();
+    let errors = assert_policy_typecheck_fails(schema_with_tags(), policy);
+    let error = assert_exactly_one_diagnostic(errors);
+    expect_err(
+        src,
+        &miette::Report::new(error),
+        &ExpectedErrorMessageBuilder::error(r#"the types String and Set<String> are not compatible"#)
+            .help("for policy `0`, both operands to a `==` expression must have compatible types. Types must be exactly equal to be compatible")
+            .exactly_one_underline(r#"principal.getTag("foo") == "bar""#)
+            .build(),
+    );
+}
+
+#[test]
+fn no_tags_allowed() {
+    // .hasTag() on an entity with no tags is allowed
+    let src = r#"
+        permit(principal, action == Action::"A4", resource) when {
+            resource.hasTag("foo")
+        };
+    "#;
+    let policy = parse_policy(Some(PolicyID::from_string("0")), src).unwrap();
+    assert_policy_typechecks(schema_with_tags(), policy);
+
+    // .getTag() on an entity with no tags is not allowed
+    let src = r#"
+        permit(principal, action == Action::"A4", resource) when {
+            resource.getTag("foo") == "bar"
+        };
+    "#;
+    let policy = parse_policy(Some(PolicyID::from_string("0")), src).unwrap();
+    let errors = assert_policy_typecheck_fails(schema_with_tags(), policy);
+    let error = assert_exactly_one_diagnostic(errors);
+    expect_err(
+        src,
+        &miette::Report::new(error),
+        &ExpectedErrorMessageBuilder::error(r#"for policy `0`, unable to guarantee safety of access to tag `"foo"` on entity type `Blank`"#)
```

Comment:
I think it's correct.  Compare that the same policy with a `has` check _does_ validate (see test immediately below), so suggesting the `has` check is a valid suggestion.  I think that `NoTagsAllowed` is actually unreachable, because it's only thrown when you do have a capability but the entity type doesn't support tags, and I don't think there's a way to have a capability when the entity type doesn't support tags, because the `.hasTag` operation would be typed `false` and prevent typechecking of the branch where you would have had the capability.

Reply:
Makes sense.

Suggesting a `has` check when that `has` would always be `false` isn't ideal and we could probably add another flag to the error to catch it (left a similar common on the `action` error message).